City-of-Helsinki/berth-reservations
# [VEN-327] | Bambora payments

## Description :sparkles:
* Add Bambora payment provider and the required flow to send payments to Bambora
* Add extra fields to models to properly work with Bambora:
  * `Order.order_number`
  * `OrderStatus.CANCELLED`
* Change `OrderLogEntry` to support more detailed satus transitions
  * `OrderLogEntry.status` -> `OrderLogEntry.to_status`
  * `OrderLogEntry.from_status`

## Issues :bug:
### Closes :no_good_woman:
**[VEN-327](https://helsinkisolutionoffice.atlassian.net/browse/VEN-327):** send order to Bambora

## Testing :alembic:
### Automated tests :gear:️
```shell
$ python manage.py payments/tests/test_bambora_payform
```

### Manual testing :construction_worker_man:
0. Set the required env variables
```
VENE_PAYMENTS_BAMBORA_API_KEY={API_KEY}
VENE_PAYMENTS_BAMBORA_API_SECRET={PRIVATE_KEY}
VENE_PAYMENTS_BAMBORA_PAYMENT_METHODS=creditcards,banks
VENE_UI_RETURN_URL=https://venepaikat.hel.fi/payment
```
1. Create a new `Order` with a `Lease` and `Application`
2. Go to the "old" api `/graphql`, and run the following mutation to get the URL:
```graphql
mutation Confirm {
  confirmPayment(input: {orderNumber: ""}) {
    url
  }
}
```
3. The mutation will return a Bambora PayForm URL, follow that URL and pay with the credit card
(if you want to try other payment methods or other cards, you can find more details on the [Bambora docs](https://payform.bambora.com/docs/web_payments/?page=testing#Test_card_numbers))
```
number = 4012888888881881
expiration = any date
CVV =  any number (3 digits)
```
4. The payment should be successful and redirect you to `https://venepaikat.hel.fi/payment?payment_status=success`

## Additional notes :spiral_notepad:
* A cron job (or something of that sort) needs to be set up to check for expired orders and change the status, running the `python manage.py expire_too_old_unpaid_orders`
* Documentation about the payment configuration and flow can be found on `docs/payments.md`
* The following environment variables have to be added:
```env
VENE_PAYMENTS_BAMBORA_API_KEY=VENE_PAYMENTS_BAMBORA_API_KEY
VENE_PAYMENTS_BAMBORA_API_SECRET=VENE_PAYMENTS_BAMBORA_API_SECRET
VENE_PAYMENTS_BAMBORA_PAYMENT_METHODS=creditcards,banks
VENE_UI_RETURN_URL=https://venepaikat.hel.fi/payments
```
